fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+use std::fs;
+use std::num::NonZeroUsize;
+use std::path::{Path, PathBuf};
+
+use data_error::{ArklibError, Result};
+use fs_atomic_light::temp_and_move;
+use fs_storage::utils::extract_key_from_file_path;
+use lru::LruCache;
+
+/// A cache entry that stores a value and its size in bytes.
+///
+/// This structure is used to track both the actual data (value)
+/// and its memory usage (size) in the cache.
+struct CacheEntry<V> {
+    value: V,
+    size: usize,
+}
+
+/// A combined in-memory and disk-based cache system.
+///
+/// This cache uses an LRU (Least Recently Used) eviction policy for the
+/// in-memory portion and persists data to disk for long-term storage.
+pub struct Cache<K, V> {
+    /// Label for logging
+    label: String,
+    /// Path to the underlying folder where data is persisted
+    path: PathBuf,
+    /// An in-memory LRU cache for quick access to frequently used items.
+    memory_cache: LruCache<K, CacheEntry<V>>,
+    /// The current memory usage in bytes.
+    current_memory_bytes: usize,
+    /// The maximum allowable memory usage in bytes.
+    max_memory_bytes: usize,
+    /// Whether to include values in debug logs
+    log_values: bool,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: Ord
+        + Clone
+        + std::fmt::Display
+        + std::hash::Hash
+        + std::str::FromStr
+        + std::fmt::Debug,
+    V: Clone + std::fmt::Debug + AsRef<[u8]> + From<Vec<u8>>,
+{
+    /// Creates a new cache instance.
+    ///
+    /// # Arguments
+    /// * `label` - Identifier used in logs
+    /// * `path` - Directory where cache files are stored
+    /// * `max_memory_bytes` - Maximum bytes to keep in memory
+    /// * `log_values` - Whether to include values in debug logs
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+        log_values: bool,
+    ) -> Result<Self> {
+        let mut cache = Self {
+            label: label.clone(),
+            path: PathBuf::from(path),
+            // TODO: NEED FIX
+            memory_cache: LruCache::new(
+                NonZeroUsize::new(max_memory_bytes)
+                    .expect("Capacity can't be zero"),
+            ),
```

Comment:
LruCache requires the capacity (number of items) to be specified during initialization. However, our Cache is designed to be limited by max_memory_bytes. So, my question is: what would be the most efficient parameter to initialize the LruCache with? In all other functions, we are already comparing based on the number of bytes, not the number of items.

I think we can create another parameter which will be `Option<usize>` with default as 100 o maybe 1000. just to initialize this LruCache everything else is fine.

Reply:
Guys, what about this? https://docs.rs/lru-mem/latest/lru_mem/

But it has only 3 stars on GitHub..

Reply:
It's actually an interesting option and would have been a perfect fit 😃

I wouldn’t recommend it though because if we find any issues later in the crate, we’d have to fork it and fix the problem ourselves, and we’re not familiar with the code. Plus, since it's not actively maintained/ used, there wouldn’t be anyone around to help us either.